metadata-io/src/test/java/com/linkedin/metadata/entity/DeleteReferencesServiceTest.java

```diff
@@ -0,0 +1,388 @@
+    final Aspect updatedAspect = DeleteReferencesService.getAspectWithReferenceRemoved("hello", aspect, schema,
+        new PathSpec("key_c", "*", "key_a"));
+
+    assertTrue(updatedAspect.data().containsKey("key_c"));
+    assertEquals(1, ((DataList) updatedAspect.data().get("key_c")).size());
+    assertTrue(((DataMap) ((DataList) updatedAspect.data().get("key_c")).get(0)).isEmpty());
+  }
+
+  /**
+   * Tests that Aspect Processor removes optional array field from struct when it is empty
+   */
+  @Test
+  public void testRemovalOptionalEmptyArray() {
+    final String value = "{\"key_c\": [{\"key_a\": \"hello\"}]}";
+    final Aspect aspect = RecordUtils.toRecordTemplate(Aspect.class, value);
+
+    final PdlSchemaParser pdlSchemaParser = new PdlSchemaParser(new DefaultDataSchemaResolver());
+    pdlSchemaParser.parse("record simple_record {\n"
+        + "key_a: string\n"
+        + "key_b: optional string\n"
+        + "}");
+
+    pdlSchemaParser.parse("record complex_record {\n"
+        + "key_c: optional array[simple_record]\n"
+        + "}");
+
+    assertEquals(1, ((DataList) aspect.data().get("key_c")).size());
+
+    final DataSchema schema = pdlSchemaParser.lookupName("complex_record");
+    final Aspect updatedAspect = DeleteReferencesService.getAspectWithReferenceRemoved("hello", aspect, schema,
+        new PathSpec("key_c", "*", "key_a"));
+
+    // contains an empty key_c
+    assertTrue(updatedAspect.data().containsKey("key_c"));
+    assertTrue(((DataList) updatedAspect.data().get("key_c")).isEmpty());
+  }
+
+  /**
+   * Tests that Aspect Processor removes nested structs more than 1 level deep from an optional field.
+   */
+  @Test
+  public void testNestedNonOptionalSubFieldsOnOptionalField() {
+    final String value = "{\"key_c\": {\"key_b\": {\"key_a\": \"hello\"}}}";
+    final Aspect aspect = RecordUtils.toRecordTemplate(Aspect.class, value);
+
+    final PdlSchemaParser pdlSchemaParser = new PdlSchemaParser(new DefaultDataSchemaResolver());
+    pdlSchemaParser.parse("record simple_record_1 {\n"
+        + "key_a: string\n"
+        + "}");
+
+    pdlSchemaParser.parse("record simple_record_2 {\n"
+        + "key_b: simple_record_1\n"
+        + "}");
+
+    pdlSchemaParser.parse("record complex_record {\n"
+        + "key_c: optional simple_record_2\n"
+        + "}");
+
+    assertTrue(aspect.data().containsKey("key_c"));
+
+    final DataSchema schema = pdlSchemaParser.lookupName("complex_record");
+    final Aspect updatedAspect = DeleteReferencesService.getAspectWithReferenceRemoved("hello", aspect, schema,
+        new PathSpec("key_c", "key_b", "key_a"));
+
+    assertFalse(updatedAspect.data().containsKey("key_c"));
+  }
+
+  /**
+   * Tests that the aspect processor is able to remove fields that are deeply nested where the top-level field is
+   * optional.
+   * This example is based on the SchemaMetadata object.
+   */
+  @Test
+  public void testRealCase() {
```

Comment:
Test "real" case? Is there something more descriptive about what is constituting a "real" case?

Reply:
testSchemaMetadataDelete -- etc